Release appcomplianceautomation microsoft.app compliance automation 2022 11 16 preview 1

Sherylueen opened the pull request:
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [x] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

Commit:
add appcomplianceautomation specs

Commit:
update appcomplianceautomation specs for 2022-11-16-preview

Sherylueen commented:
All the approved RPSaasMaster PRs:
**https://github.com/Azure/azure-rest-api-specs-pr/pull/7967**
https://github.com/Azure/azure-rest-api-specs-pr/pull/8103
https://github.com/Azure/azure-rest-api-specs-pr/pull/8219
**https://github.com/Azure/azure-rest-api-specs-pr/pull/8653
https://github.com/Azure/azure-rest-api-specs-pr/pull/8714
https://github.com/Azure/azure-rest-api-specs-pr/pull/8778**
https://github.com/Azure/azure-rest-api-specs-pr/pull/9196
https://github.com/Azure/azure-rest-api-specs-pr/pull/9206
All the bold prs are ARMSignedOff

Sherylueen commented:
> The api-version included supposed to be approved in RPSaaSMaster.
> 
> However I've no idea why we use -beta or -privatepreview as api-version.
> 
> Hence requesting sign-off from ARM.

Invite @suhasaraos as a reviewer. I saw there are some -privatepreview version in public repo. But -beta is very rare.
So I guess maybe -privatepreview is reasonable, and we can keep beta in the private repo.
Looking forward to Suhas's opinion.

raosuhas commented:
> > The api-version included supposed to be approved in RPSaaSMaster.
> > However I've no idea why we use -beta or -privatepreview as api-version.
> > Hence requesting sign-off from ARM.
> 
> Invite @suhasaraos as a reviewer. I saw there are some -privatepreview version in public repo. But -beta is very rare. So I guess maybe -privatepreview is reasonable, and we can keep beta in the private repo. Looking forward to Suhas's opinion.

Hi @Sherylueen 

Can you please create a separate PR per api version ? What are you trying to achieve here ? Are these versions already signed off in RPSaaSMaster and is there any diff between the version checked into the private repo and here ?

Sherylueen committed:
remove privatepreview and beta (#21124)

Co-authored-by: Yanwen Liu <yanwenliu@microsoft.com>

Sherylueen commented:
According to the feedback from ARM. I should not add any suffix other than -preview.
So I removed -beta and -privatepreview  versions, after that I removed ARMChangesRequested label.

Sherylueen commented:
Hi @raosuhas ~
I removed the beta and privatepreview versions, just keep preview now.
Please have a look, when you got time.

raosuhas commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "OK. The request has succeeded."
```
Dont you need an actual response body here for the operation result itself ?

qinqingxu replied:
@raosuhas, This is a historical API, at the very begginning, we were plan to add an API to return the operation status with a body. But finally we turn to using the RPaaS. So that not support this API. Maybe we can think about how to mitigate it. Maybe we can remove it.

Sherylueen replied:
Hi @raosuhas ~
This historical API has been removed. See [here](https://github.com/Azure/azure-rest-api-specs/pull/21152/files#diff-16cc0066db31d47ee8b4f0e9385ece034881677d74a14b4d48790444b24bda4d).

raosuhas commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+          "202": {
+            "description": "Accepted. The request is still running."
+          },
+          "204": {
```
I dont think you need the 204. Id a resource is not present then the delete should fail immediately. No need to start a long running operation.

qinqingxu replied:
@raosuhas Yes, we can remove the Long-running operation for the GET operation.

raosuhas commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+        "x-ms-long-running-operation": true,
+        "x-ms-long-running-operation-options": {
+          "final-state-via": "azure-async-operation"
+        },
```
Get operation status should not be a long running operation

qinqingxu replied:
@raosuhas good suggestion.

Sherylueen replied:
API removed.

raosuhas commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+        "tags": [
+          "AppComplianceAutomation"
+        ],
+        "description": "Get the AppComplianceAutomation report list.",
```
>Get the AppComplianceAutomation report list. [](http://example.com/codeflow?start=24&length=44)

nit : 

Get the AppComplianceAutomation report list for the tenant

qinqingxu replied:
@raosuhas Do you mean, we need to highlight the resource is base on Subscription level, or base on the tenant level?

Sherylueen replied:
Hi @raosuhas ~
This description has been changed to "Get the AppComplianceAutomation report list for the tenant".
See [here](https://github.com/Azure/azure-rest-api-specs/pull/21152/files#diff-16cc0066db31d47ee8b4f0e9385ece034881677d74a14b4d48790444b24bda4dR81)

raosuhas commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+            "description": "Long running operation support.",
+            "schema": {
+              "$ref": "#/definitions/ReportResource"
+            }
```
Long running PATCH should return 202 : 

[https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#updating-using-patch](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#updating-using-patch)

qinqingxu replied:
@raosuhas Yes, we implemented the code which will return 202. We can fix it in the API specs.

Sherylueen replied:
Hi @raosuhas ~
The statusCode has been changed from 201 to 202. See [here](https://github.com/Azure/azure-rest-api-specs/pull/21152/files#diff-16cc0066db31d47ee8b4f0e9385ece034881677d74a14b4d48790444b24bda4dR245).

Sherylueen replied:
After testing we found that RPSaas would change our 202 statuscode to 201. See [here.](https://stackoverflow.microsoft.com/questions/311423)
So we keep both 202 and 201 in the spec.

raosuhas commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "Success. The response indicates the resource is already deleted."
```
>he response indicates the resource is already deleted. [](http://example.com/codeflow?start=38&length=54)

nit : update description.  "... the resource has been deleted" to differentiate from the 204 result

qinqingxu replied:
@raosuhas Yes, thanks.

Sherylueen replied:
Hi @raosuhas ~
Description updated, please see [here](https://github.com/Azure/azure-rest-api-specs/pull/21152/files#diff-16cc0066db31d47ee8b4f0e9385ece034881677d74a14b4d48790444b24bda4dR281).

raosuhas commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+          },
+          {
+            "$ref": "#/parameters/OfferGuidParameter"
+          }
```
POST parameters should be in the body not in the query 

[https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#resource-action-requests](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#resource-action-requests)

qinqingxu replied:
@raosuhas

Thanks for your link.
But I have few questions.

1. Actually I saw there are a lot of RP, their API specs also using the query parameter in their POST action.
2. In this [Microsoft API guidance](https://microsoft.sharepoint.com/:p:/t/AzureAPIStewardshipBoard/EQZVvzdBQqlGoddNjiEc_jkB61r5G6HuDX03oum0254z6w?e=7l5uME). It also give a lot of sample of using query parameter for the POST actions.

Sherylueen replied:
Hi @raosuhas ~
We moved some parameters to body now. only keep the in path parameters outside.
See [here](https://github.com/Azure/azure-rest-api-specs/pull/21152/files#diff-16cc0066db31d47ee8b4f0e9385ece034881677d74a14b4d48790444b24bda4dL467).

raosuhas commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+      "description": "The overview of the compliance result for one report.",
+      "type": "object",
+      "properties": {
+        "passed": {
```
>passed [](http://example.com/codeflow?start=9&length=6)

nit : consider being more descriptive

passedCount , failedCount, etc

qinqingxu replied:
@raosuhas Thanks for your comments. We can think about how to migrate this variable name.

Sherylueen replied:
Hi @raosuhas ~
Fixed! Please review [here](https://github.com/Azure/azure-rest-api-specs/pull/21152/files#diff-16cc0066db31d47ee8b4f0e9385ece034881677d74a14b4d48790444b24bda4dL519).

raosuhas commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+        "isPass": {
+          "type": "boolean",
+          "readOnly": true,
+          "description": "Indicates whether all the resource(s) are compliant."
```
consider updating to enum  [https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html#common-issues-found-in-review](https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html#common-issues-found-in-review)

Replace boolean/switch properties with better enum
A Boolean will forever have two valid values (true or false). A string enum type is always preferred. Also, properties should always provide better values just than True and False. For example two switches "isTypeA" and "isTypeB" should be replaced with one enum "type": [A, B, DefaultType]. Even if you still believe [True, False] are the correct values for a property, you should use a string enum with values [True, False] instead of boolean. Enums are always a more flexible and future proof option because they allow additional values to be added in the future in a non-breaking way, e.g. [True, False, Unknown].

qinqingxu replied:
@raosuhas Do you mean change to like "true", "false" string enum?

Sherylueen replied:
Hi @raosuhas ~
Fixed, please review [here](https://github.com/Azure/azure-rest-api-specs/pull/21152/files#diff-16cc0066db31d47ee8b4f0e9385ece034881677d74a14b4d48790444b24bda4dL838).

raosuhas commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/examples/Report_CreateOrUpdate.json:
```diff
@@ -0,0 +1,117 @@
+            "resourceId": "/subscriptions/00000000-0000-0000-0000-000000000000/resourcegroups/myResourceGroup/providers/Microsoft.Network/privateEndpoints/myPrivateEndpoint",
+            "tags": {
+              "key1": "value1"
+            }
```
are thee actual tags of the ARM resource above ? Will this operation actually update them ?

qinqingxu replied:
@raosuhas, Yes, these tags are the ARM resources' tag. But we just store their info, and not update them.

raosuhas commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+            "sasUri": {
+              "description": "uri of compliance detailed pdf report",
+              "type": "string",
+              "readOnly": true
```
could this contain secrets in the query string like sig values etc ? if so please mark as x-ms-secret

qinqingxu replied:
@raosuhas Yes, I think we can add the "**x-ms-secret**".

Sherylueen replied:
Hi @raosuhas ~
Actually the sasUri is not a secret, we don't get the value from users' input. Our service generate the sasUri, and return the value in the response.
When user get the sasUri, they can use the sasUri to download a file within 10 minutes. After 10 minutes, the uri will expire.
So add a "x-ms-secret", but set the value to false temporarily.
See [here](https://github.com/Azure/azure-rest-api-specs/pull/21152/files#diff-16cc0066db31d47ee8b4f0e9385ece034881677d74a14b4d48790444b24bda4dL1183).

TimLovellSmith replied:
A sasUri is a secret, even when it is not from the user's input, because it is a secret value that allows a privileged operation, in this case, downloading a file. You should be able to mark it as x-ms-secret as long as it is only returned from POST requests, and not GET requests.

qinqingxu commented:
@raosuhas Regarding to all of your comments.
Some of them, we can directly to modify, and it will not impact our current code implementation at all.
But some of modification might impact the code implementation.
And the fact is: Once we change the API spec, and we need to do the rollout. And during the rollout done, we must immediately to change our code for the new API spec.

And that will be a gap, which might cause the service down.

So, I have an idea.

First, we can modify the "harmless" comments, like: Add the "x-ms-secret", Remove the "Long-running" tag....
And regarding to other comments which might caused the code change. Maybe we can postpone to the next API version?

Current our API version will be published at the Nov-16, after that we will create a new API version for the GA target.

Sherylueen committed:
Release appcomplianceautomation microsoft.app compliance automation 2022 11 16 preview resolve comments 2 (#21152)

* fix comments

* remove duplicate in body

* resolve comments

* edit examples to match schema

* fix failure

* fix model validation

* change passed to passedCount

* add x-ms-enum to isPass

* change delete 200 description

Co-authored-by: Yanwen Liu <yanwenliu@microsoft.com>

Sherylueen committed:
set sasUri x-ms-secret to true (#21161)

Co-authored-by: Yanwen Liu <yanwenliu@microsoft.com>

jianyexi commented:
/azp run

Sherylueen committed:
remove properties from download request body (#21184)

* remove properties from download request body

* remove properties in snapshot_resourcelist_download.json

* keep statuscode 201 and 202 in patch report response, due to RPSaas will change 202 to 201

* add body in patch report 201 response

Co-authored-by: Yanwen Liu <yanwenliu@microsoft.com>

TimLovellSmith commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+      "enum": [
+        "Succeeded",
+        "Failed",
+        "Cancelled",
```
>Cancelled [](http://example.com/codeflow?start=9&length=9)

I hear that provisioning state has to be spelled 'Canceled' with only one letter 'l' (L) to be properly recognized by ARM.

qinqingxu replied:
@TimLovellSmith  Do you have the reference link for this recommendation?

Sherylueen replied:
Fixed, "Cancelled" has been changed to "Canceled".

TimLovellSmith commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+        },
+        "resourceId": {
+          "type": "string",
+          "description": "The compliance result mapped resource Id.",
```
>       "description": "The compliance result mapped resource Id.", [](http://example.com/codeflow?start=3&length=66)

Should this be format arm-id? Probably doesn't matter much as long as its readonly.

qinqingxu replied:
Yes, it is for the arm resource id, like /subscriptions/{subscriptionId}/resourceGroup/{rg_name}/{RP namespace}/.....

Sherylueen replied:
Fixed, example added.
Thanks Tim~

TimLovellSmith commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+        },
+        "controlId": {
+          "type": "string",
+          "description": "The control Id.",
```
>ion": "The control Id.", [](http://example.com/codeflow?start=19&length=24)

If your docs say what format is it in, it helps people understand how to actually use the api

Sherylueen replied:
Fixed, thanks Tim.

TimLovellSmith commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+        },
+        "complianceState": {
+          "type": "string",
+          "description": "The compliance result's status.",
```
>description [](http://example.com/codeflow?start=11&length=11)

Is there an enum of known values? Or is it freeform?

Sherylueen replied:
Yes, we have changed complianceState to enum, the available values can be "Healthy" and "Unhealthy".
Thanks Tim~

TimLovellSmith commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+        },
+        "statusChangeDate": {
+          "description": "The status change date for the resource. For unavailable date, set it as N/A.",
+          "type": "string",
```
Would this be date-time format? What does 'set it as N/A' mean when its readonly, do you mean the server will be returning the string 'N/A'? I wonder if that makes it easier or harder to parse than just returning null/undefined when its unknown.

TimLovellSmith commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+    },
+    "ReportResource": {
+      "type": "object",
+      "description": "A class represent a AppComplianceAutomation report resource.",
```
>a [](http://example.com/codeflow?start=40&length=1)

Should be 'an AppComplianceAutomation...', not 'a'

Sherylueen replied:
Fixed.

TimLovellSmith commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+        },
+        "timeZone": {
+          "type": "string",
+          "description": "Report collection trigger time's time zone."
```
>n trigger time's time zone." [](http://example.com/codeflow?start=42&length=28)

Document the set of valid values?

Sherylueen replied:
Fixed.

TimLovellSmith commented on specification/appcomplianceautomation/resource-manager/Microsoft.AppComplianceAutomation/preview/2022-11-16-preview/appcomplianceautomation.json:
```diff
@@ -0,0 +1,1374 @@
+        }
+      }
+    },
+    "ReportProperties": {
```
>ReportProperties [](http://example.com/codeflow?start=5&length=16)

if any of these are not alterable using PATCH, use x-ms-mutability

Sherylueen replied:
Hi Tim,
Temporarily the PUT and PATCH alterable scopes are the same.
So we only use "readonly" to mark unmodifiable variables.
Thanks for the comments, we will reconsider the specs in next version.

Sherylueen committed:
change ProvisioningState Cancelled to Canceled, and some descriptions (#21237)

* change ProvisioningState Cancelled to Canceled, and some descriptions

* add x-ms-enum extension to ComplianceState

Co-authored-by: Yanwen Liu <yanwenliu@microsoft.com>
